### DISCUSSION TITLE
Configure PyCharm for pyscript development

### DISCUSSION THREAD

**Discussion by Gvolten** [custom-components/pyscript#566] [Q&A]
When I try to use PyCharm to edit pyscript files, I obviously get unresolved reference errors. Is there any way to import pyscript into PyCharm such that it recognizes all of the references?

**Comment by ALERTua** [comment]
Here's your answer: https://github.com/dmamelin/pyscript_autocomplete

**Reply by Gvolten** [reply]
How can I use it if I have configured pyscript using the GUI?
Why is allowing all imports necessary?

**Reply by ALERTua** [reply]
> How can I use it if I have configured pyscript using the GUI?

By following the installation instructions :)


> Why is allowing all imports necessary?

Because pyscript_autocomplete creates a module that is being imported by all of your scripts. The module that describes all the builtins and all of your services.

**Reply by ghost** [reply]
@ALERTua 
Very grateful. Without pyscript_autocomplete, using pyscript causes too much stress. Did you try to suggest that the author integrate pyscript_autocomplete into pyscript? It would be very cool

**Reply by ALERTua** [reply]
> @ALERTua Very grateful. Without pyscript_autocomplete, using pyscript causes too much stress. Did you try to suggest that the author integrate pyscript_autocomplete into pyscript? It would be very cool

that's for @craigbarratt to decide :)

**Reply by ghost** [reply]
@craigbarratt do you have any plans to add something similar to the creation of @dmamelin   (or simply integrate his [pyscript_autocomplete](https://github.com/dmamelin/pyscript_autocomplete))?

[hass-pyscript-jupyter](https://github.com/craigbarratt/hass-pyscript-jupyter/issues) is [broken](https://github.com/craigbarratt/hass-pyscript-jupyter/issues) and has not been supported for a long time. And writing with autocompletion in IDE is much more pleasant than in notebook.

(it would be a dream to autocomplete with type hints)

**Reply by dmamelin** [reply]
@craigbarratt  I can rework pyscript_autocomplete and make a PR

**Reply by ghost** [reply]
> @craigbarratt I can rework pyscript_autocomplete and make a PR

@dmamelin By the way, do you think it would be better to turn mocks into [stubs](https://mypy.readthedocs.io/en/stable/stubs.html) for the basic API? Then instead of weird hacks it will just be .pyi file in [typeshed](https://github.com/python/typeshed)

**Reply by dmamelin** [reply]
Thanks, I will take a look. I'm new to python

**Reply by esauvisky** [reply]
Thank you so much for this, terrific job for someone new at python! Honestly, I wouldn't even be trying out pyscript if it wasn't for this. This should definitely be a core component, particularly because the jupyter thing is not at all simple, I don't personally find it developer-friendly, and it doesn't _actually_ offers proper type annotations.

Also, there's a lot of cool things that could be done by integrating this into pyscript, like dynamically generating .pyi stub files as entities change, eliminating the need for importing mocks. If someone takes the initiative of making this happen, I'll contribute for sure.

**Reply by mbnguyen** [reply]
This is genius! Thank you so much @dmamelin! I personally can't stand the Jupyter approach.
